### DISCUSSION TITLE
Help in understanding how lambda works for page redirects

### DISCUSSION THREAD

**Discussion by Anindya088** [zauberzeug/nicegui#1943] [Q&A] [answered]
### Question

Hi All,

Can you please help me what is the purpose/use of lambda functions and what is exactly happening for the page redirects when we use buttons to go from one page to another?

Example sample code below:
```python
from nicegui import ui

@ui.page('/pageA')
def pageA():
    ui.label('This is page A')
    ui.button('Go to Page B', on_click=lambda: ui.open('/pageB'))

@ui.page('/pageB')
def pageB():
    ui.label('This is page B')
    ui.button('Go to Page A', on_click=lambda: ui.open('/pageA'))

ui.run()
```
So if I don't use the ```lambda``` in the ```on_click```, then I get an error message in console saying
```Cannot open page because client is not connected, try RedirectResponse from FastAPI instead```
But using the lamda, the page redirects work just fine. 

Not that this is any issue or anything, but I am just curious to understand, what is really happening with this lambda here? Why does it work when lambda is there but does not when removing it?

Thanks a lot,
Anindya

**Comment by ed2050** [comment, marked as answer]
Because on_click takes a function argument.  You pass the name of a function, and on_click calls that function later, when the button is actually clicked.

`ui.open` is a reference to a function.  `ui.open ('/pageB')` is a call to that function.  Whatever value that function returns will be used at the point in your code.  Eg `mydict [foo] = ui.open ('pageB')` will put the return value in mydict [foo].

The lambda changes `ui.open ('/pageB')` from a function call to a function reference.  lambda defines a new function and provides a reference to it.  You pass that function reference to on_click.  `ui.button` saves the function reference and calls it later when button is clicked.

So `ui.button (on_click = ui.page ('/pageB')` calls `ui.page ('/pageB')` at that point in your code.  Its return value will then be passed as the on_click parameter.  Since ui.run hasn't started yet, your app isn't running, there's no client connected, and `ui.page ()` can't send the client to pageB.  That's why you see the error message.

In theory, you could set `ui.button (on_click = ui.page)` to have your button call ui.page when the button is clicked.  However ui.button calls its on_click func with no args [1]; it doesn't know to pass 'pageA' or 'pageB' or anything else.  It would just call `ui.page()`, which doesn't do what you want.  Using the lambda creates a function with no args for ui.button to call later.  The expression inside the lambda then calls ui.page with the 'pageA' arg to send users to the page you want.  With some closure magic to save the 'pageA' arg for later.

You can read more about how lambdas work and when to use them here:
https://realpython.com/python-lambda/

[1] it seems that technically nicegui can pass event info as args to on_click handler.  [See here](https://nicegui.io/documentation/button).  But that won't have the 'pageA' or 'pageB' you need.  How nicegui knows not to pass event args to a no-arg lambda, I'm not sure.  Maybe `inspect` to figure out  how many args the passed on_click function accepts.

HTH,

**Reply by Anindya088** [reply]
Hi @ed2050 , 

Thanks so much for the detailed explanation, this is very clear and very helpful. I did not realize that python will run the code sequentially and hence the page call is happening before the ```ui.run``` which cannot be. Okay, so ```lambda``` is a basically a nice way to change that into a reference to be called later, as you explained.

Thanks a lot again,
Anindya
